### PR TITLE
add cwtool to flake packages.default

### PR DESCRIPTION
**Pull request by chessai**
I would prefer if this was built by default with `nix build`

### CODE REVIEW

**Review by imalsogreg — APPROVED** [review]
After this, `chainweb-node` and `cwtool` are both available under `result` when building the flake's default package:

```
[greghale@michelangelo:~/code/chainweb-node]$ ls result/bin
chainweb-node  cwtool
```

👍  Looks great!

**Comment by enobayram** [comment]
Hi @chessai Thanks for all the improvements you've made to the Nix files in this PR in addition to adding `cwtool` to the `default` package. I have no objection to this PR structurally, but I would like to point out the increase in closure size:

Here's the closure size from this PR:
```
$ nix path-info --closure-size $(nix build github:kadena-io/chainweb-node?rev=a3fdfe4350b3e1a243e5db3a4442681418fca10b --print-out-paths)
/nix/store/4c4k3j2vkfsgzc07h6jv1c2gjmlpszfg-chainweb	  757657432
```

And here's the closure size from before:
```
$ nix path-info --closure-size $(nix build github:kadena-io/chainweb-node?rev=4443f711256e575b008896c49b42626cde2e1463 --print-out-paths)
/nix/store/w05ff2sa5n01ahp4h3byng4blpvfaw4a-chainweb-exe-chainweb-node-2.18.1	  510178600
```
So, we see ~250MB (50%) increase in the closure size for the default package. This may or may not be an acceptable compromise depending on how often people need `cwtool`. One benefit of nix-flakification was that it's now so much easier to access these outputs even when they're not in the default package, so we could consider exposing `cwtool` as as `packages.cwtool` so that anyone can just fetch it by running `nix build github:kadena-io/chainweb-node#cwtool`.

If the intention of this PR is just to make sure that `cwtool` gets built as part of CI and not necessarily becomes a part of the `default` package, then we could just add it to [the `check` package](https://github.com/kadena-io/chainweb-node/blob/d06e8974b1e9b8dd32044f6843d9ac4e5b00fecb/flake.nix#L36) that the GitHub nix build action builds.

---------------------

On a tangential note, I think the current 510MB closure size for the default package doesn't make much sense either. See what's in the closure of the chainweb-node binary:
```
$ nix-store -q --requisites $(nix build github:kadena-io/chainweb-node?rev=4443f711256e575b008896c49b42626cde2e1463 --print-out-paths)
/nix/store/34xlpp3j3vy7ksn09zh44f1c04w77khf-libunistring-1.0
/nix/store/5mh5019jigj0k14rdnjam1xwk5avn1id-libidn2-2.3.2
/nix/store/4nlgxhb09sdr51nc9hdm8az5b08vzkgx-glibc-2.35-163
/nix/store/026hln0aq1hyshaxsdvhg0kmcm6yf45r-zlib-1.2.13
/nix/store/2rgwrp4zwia0ayhhxzw000cnjvr51pxa-lz4-1.9.4
/nix/store/akqx5jv5ax8rbjyi72h5644q00kdpjdy-lz4-1.9.4-bin
/nix/store/0rc6yf1qpl09xz4k35xcisaq9mmv5v58-lz4-1.9.4-dev
/nix/store/0y971q54v6jm9ss243xhl4y0gnlsm9c8-zlib-1.2.13-dev
/nix/store/1dgws25664p544znpc6f1nh9xmjf4ykc-pcre-8.45
/nix/store/57xv61c5zi8pphjbcwxxjlgc34p61ic9-glibc-2.35-163-bin
/nix/store/i38jcxrwa4fxk2b7acxircpi399kyixw-linux-headers-6.0
/nix/store/4pqv2mwdn88h7xvsm7a5zplrd8sxzvw0-glibc-2.35-163-dev
/nix/store/mdck89nsfisflwjv6xv8ydj7dj0sj2pn-gcc-11.3.0-lib
/nix/store/1gf2flfqnpqbr1b4p4qz2f72y42bs56r-gcc-11.3.0
/nix/store/xghl0l2f6byzqvy06ana2y4isqwa7blw-libffi-3.4.4
/nix/store/40q4166vs7c5v5qx2si6b5fdvjml3p5p-libffi-3.4.4-dev
/nix/store/4mxnw95jcm5a27qk60z7yc0gvxp42b9a-openssl-3.0.7
/nix/store/5056r5jmd1p22ji9wlh24gdj2zvhggfm-snappy-1.1.9
/nix/store/664hlv5rfi4j1z1sa6jzj49sg7l2k03m-numactl-2.0.16
/nix/store/dsd5gz46hdbdk2rfdimqddhq6m8m8fqs-bash-5.1-p16
/nix/store/70in66z4smj2kxs24jibnrj3r5q9gryp-gflags-2.2.2
/nix/store/86bp03jkmsl6f92w0yzg4s59g5mhxwmy-gnugrep-3.7
/nix/store/w10in9diaqrcqqxi5lg20n3q2jfpk6pq-zstd-1.5.2
/nix/store/mb0pcxkmrg0f6k0zaywlnvpk9q3j5ans-zstd-1.5.2-bin
/nix/store/90jpxayjm6an13vxnpcyvj6jdnjivg1r-zstd-1.5.2-dev
/nix/store/ysl6qj5r7nn63b16954dhk7x47r5yq7i-bzip2-1.0.8
/nix/store/a8mhcagrsly7c7mpjrpsnaahk4aax056-bzip2-1.0.8-bin
/nix/store/972ynh8bnpd65s7cb23x21d2kvc46yfy-bzip2-1.0.8-dev
/nix/store/i2gm5vqgvlrxbjnj8jy3aw29mgfb0nr7-snappy-1.1.9-dev
/nix/store/ijz81p08bp812q2bvv77lz9qpfzncibd-gmp-with-cxx-6.2.1
/nix/store/w05ff2sa5n01ahp4h3byng4blpvfaw4a-chainweb-exe-chainweb-node-2.18.1
```

There's probably a lot of room for improvement here.

**Review by enobayram — APPROVED** [review]
This is a nice PR that comes with a closure size trade-off (see [my comment](https://github.com/kadena-io/chainweb-node/pull/1642#issuecomment-1534302283)). Thanks for all the improvements. I'm approve this PR since it looks fine if the trade-off is deemed acceptable.

**Review by jwiegley — APPROVED** [review]
@enobayram `cwtool` is pretty essential, and is well worth the 250 MB closure size increase.

**Comment by chessai** [comment]
@enobayram I actually was really interested in why the closure size was so large.

TLDR got it down to 64MB

Not TLDR debugging train of thought:
```
❯ du -hc --max=0 $(nix-store -qR ./result) | sort -h
2.0K    /nix/store/4c4k3j2vkfsgzc07h6jv1c2gjmlpszfg-chainweb
16K    /nix/store/972ynh8bnpd65s7cb23x21d2kvc46yfy-bzip2-1.0.8-dev
21K    /nix/store/40q4166vs7c5v5qx2si6b5fdvjml3p5p-libffi-3.4.4-dev
35K    /nix/store/5056r5jmd1p22ji9wlh24gdj2zvhggfm-snappy-1.1.9
40K    /nix/store/xghl0l2f6byzqvy06ana2y4isqwa7blw-libffi-3.4.4
57K    /nix/store/i2gm5vqgvlrxbjnj8jy3aw29mgfb0nr7-snappy-1.1.9-dev
60K    /nix/store/ysl6qj5r7nn63b16954dhk7x47r5yq7i-bzip2-1.0.8
63K    /nix/store/a8mhcagrsly7c7mpjrpsnaahk4aax056-bzip2-1.0.8-bin
65K    /nix/store/0rc6yf1qpl09xz4k35xcisaq9mmv5v58-lz4-1.9.4-dev
65K    /nix/store/0y971q54v6jm9ss243xhl4y0gnlsm9c8-zlib-1.2.13-dev
89K    /nix/store/026hln0aq1hyshaxsdvhg0kmcm6yf45r-zlib-1.2.13
128K    /nix/store/90jpxayjm6an13vxnpcyvj6jdnjivg1r-zstd-1.5.2-dev
155K    /nix/store/2rgwrp4zwia0ayhhxzw000cnjvr51pxa-lz4-1.9.4
219K    /nix/store/akqx5jv5ax8rbjyi72h5644q00kdpjdy-lz4-1.9.4-bin
221K    /nix/store/mb0pcxkmrg0f6k0zaywlnvpk9q3j5ans-zstd-1.5.2-bin
232K    /nix/store/664hlv5rfi4j1z1sa6jzj49sg7l2k03m-numactl-2.0.16
271K    /nix/store/5mh5019jigj0k14rdnjam1xwk5avn1id-libidn2-2.3.2
313K    /nix/store/1dgws25664p544znpc6f1nh9xmjf4ykc-pcre-8.45
493K    /nix/store/ijz81p08bp812q2bvv77lz9qpfzncibd-gmp-with-cxx-6.2.1
610K    /nix/store/70in66z4smj2kxs24jibnrj3r5q9gryp-gflags-2.2.2
631K    /nix/store/86bp03jkmsl6f92w0yzg4s59g5mhxwmy-gnugrep-3.7
634K    /nix/store/w10in9diaqrcqqxi5lg20n3q2jfpk6pq-zstd-1.5.2
920K    /nix/store/dsd5gz46hdbdk2rfdimqddhq6m8m8fqs-bash-5.1-p16
983K    /nix/store/34xlpp3j3vy7ksn09zh44f1c04w77khf-libunistring-1.0
1.9M    /nix/store/57xv61c5zi8pphjbcwxxjlgc34p61ic9-glibc-2.35-163-bin
2.5M    /nix/store/4pqv2mwdn88h7xvsm7a5zplrd8sxzvw0-glibc-2.35-163-dev
3.6M    /nix/store/4mxnw95jcm5a27qk60z7yc0gvxp42b9a-openssl-3.0.7
4.1M    /nix/store/mdck89nsfisflwjv6xv8ydj7dj0sj2pn-gcc-11.3.0-lib
5.9M    /nix/store/i38jcxrwa4fxk2b7acxircpi399kyixw-linux-headers-6.0
18M    /nix/store/4nlgxhb09sdr51nc9hdm8az5b08vzkgx-glibc-2.35-163
107M    /nix/store/1gf2flfqnpqbr1b4p4qz2f72y42bs56r-gcc-11.3.0
132M    /nix/store/480clkf74y2fqw8mjvp6vnz4sz6vzwx2-chainweb-exe-cwtool-2.18.1
132M    /nix/store/b82prm0kc08fs4izr2n9db4b41fsk1qg-chainweb-exe-chainweb-node-2.18.1
412M    total
```
You'll notice that gcc is in there. Strange...

I ran a bunch of `nix why-depends` and `grep` commands, but they showed nothing useful:
```
❯ nix why-depends ./result /nix/store/1gf2flfqnpqbr1b4p4qz2f72y42bs56r-gcc-11.3.0
/nix/store/4c4k3j2vkfsgzc07h6jv1c2gjmlpszfg-chainweb
└───/nix/store/480clkf74y2fqw8mjvp6vnz4sz6vzwx2-chainweb-exe-cwtool-2.18.1
    └───/nix/store/1gf2flfqnpqbr1b4p4qz2f72y42bs56r-gcc-11.3.0

❯ grep -r 480clkf74y2fqw8mjvp6vnz4sz6vzwx2 /nix/store/4c4k3j2vkfsgzc07h6jv1c2gjmlpszfg-chainweb

❯ grep -r 1gf2flfqnpqbr1b4p4qz2f72y42bs56r /nix/store/480clkf74y2fqw8mjvp6vnz4sz6vzwx2-chainweb-exe-cwtool-2.18.1

❯ ls -lR /nix/store/4c4k3j2vkfsgzc07h6jv1c2gjmlpszfg-chainweb | grep 480clkf74y2fqw8mjvp6vnz4sz6vzwx2
lrwxrwxrwx 81 root 31 Dec  1969  cwtool -> /nix/store/480clkf74y2fqw8mjvp6vnz4sz6vzwx2-chainweb-exe-cwtool-2.18.1/bin/cwtool

❯ ls -lR /nix/store/480clkf74y2fqw8mjvp6vnz4sz6vzwx2-chainweb-exe-cwtool-2.18.1 | grep 1gf2flfqnpqbr1b4p4qz2f72y42bs56r
```
As expected, chainweb has a symlink that depends on cwtool, but at this point it's not clear why cwtool depends on gcc

GCC is a problem, but gcc-lib isn't really (huge size difference):
```
4.1M    /nix/store/mdck89nsfisflwjv6xv8ydj7dj0sj2pn-gcc-11.3.0-lib
107M    /nix/store/1gf2flfqnpqbr1b4p4qz2f72y42bs56r-gcc-11.3.0
```

`nix why-depends --help` reveals `--precise`:
```
❯ nix why-depends ./result /nix/store/1gf2flfqnpqbr1b4p4qz2f72y42bs56r-gcc-11.3.0 --precise
/nix/store/4c4k3j2vkfsgzc07h6jv1c2gjmlpszfg-chainweb
└───bin/cwtool -> /nix/store/480clkf74y2fqw8mjvp6vnz4sz6vzwx2-chainweb-exe-cwtool-2.18.1/bin/cwtool
    → /nix/store/480clkf74y2fqw8mjvp6vnz4sz6vzwx2-chainweb-exe-cwtool-2.18.1
    └───bin/cwtool: …29.3/include/rocksdb./nix/store/1gf2flfqnpqbr1b4p4qz2f72y42bs56r-gcc-11.3.0/include/>
        → /nix/store/1gf2flfqnpqbr1b4p4qz2f72y42bs56r-gcc-11.3.0
```
(Don't know why grep didn't see this, but oh well)

So it looks like the debug info for cwtool includes the include paths for compiled c code (mostly rocksdb?)
And at this point I think that nix assumes that means, you need all of gcc at runtime

Now I'm thinking, I should just be able to run `strip` and be fine. So I replace `default` with:
```nix
    default = pkgs.runCommandCC "chainweb" {} ''
      mkdir -pv $out/bin
      cp ${flake.packages."chainweb:exe:chainweb-node"}/bin/chainweb-node $out/bin/chainweb-node
      cp ${flake.packages."chainweb:exe:cwtool"}/bin/cwtool $out/bin/cwtool
      chmod +w $out/bin/{cwtool,chainweb-node}
      $STRIP $out/bin/chainweb-node
      $STRIP $out/bin/cwtool
    '';
```

But this doesn't really make much of a dent:
```
❯ du -hc --max=0 $(nix-store -qR ./result) | sort -h
16K    /nix/store/972ynh8bnpd65s7cb23x21d2kvc46yfy-bzip2-1.0.8-dev
21K    /nix/store/40q4166vs7c5v5qx2si6b5fdvjml3p5p-libffi-3.4.4-dev
35K    /nix/store/5056r5jmd1p22ji9wlh24gdj2zvhggfm-snappy-1.1.9
40K    /nix/store/xghl0l2f6byzqvy06ana2y4isqwa7blw-libffi-3.4.4
57K    /nix/store/i2gm5vqgvlrxbjnj8jy3aw29mgfb0nr7-snappy-1.1.9-dev
60K    /nix/store/ysl6qj5r7nn63b16954dhk7x47r5yq7i-bzip2-1.0.8
63K    /nix/store/a8mhcagrsly7c7mpjrpsnaahk4aax056-bzip2-1.0.8-bin
65K    /nix/store/0rc6yf1qpl09xz4k35xcisaq9mmv5v58-lz4-1.9.4-dev
65K    /nix/store/0y971q54v6jm9ss243xhl4y0gnlsm9c8-zlib-1.2.13-dev
89K    /nix/store/026hln0aq1hyshaxsdvhg0kmcm6yf45r-zlib-1.2.13
128K    /nix/store/90jpxayjm6an13vxnpcyvj6jdnjivg1r-zstd-1.5.2-dev
155K    /nix/store/2rgwrp4zwia0ayhhxzw000cnjvr51pxa-lz4-1.9.4
219K    /nix/store/akqx5jv5ax8rbjyi72h5644q00kdpjdy-lz4-1.9.4-bin
221K    /nix/store/mb0pcxkmrg0f6k0zaywlnvpk9q3j5ans-zstd-1.5.2-bin
232K    /nix/store/664hlv5rfi4j1z1sa6jzj49sg7l2k03m-numactl-2.0.16
271K    /nix/store/5mh5019jigj0k14rdnjam1xwk5avn1id-libidn2-2.3.2
313K    /nix/store/1dgws25664p544znpc6f1nh9xmjf4ykc-pcre-8.45
493K    /nix/store/ijz81p08bp812q2bvv77lz9qpfzncibd-gmp-with-cxx-6.2.1
610K    /nix/store/70in66z4smj2kxs24jibnrj3r5q9gryp-gflags-2.2.2
631K    /nix/store/86bp03jkmsl6f92w0yzg4s59g5mhxwmy-gnugrep-3.7
634K    /nix/store/w10in9diaqrcqqxi5lg20n3q2jfpk6pq-zstd-1.5.2
920K    /nix/store/dsd5gz46hdbdk2rfdimqddhq6m8m8fqs-bash-5.1-p16
983K    /nix/store/34xlpp3j3vy7ksn09zh44f1c04w77khf-libunistring-1.0
1.9M    /nix/store/57xv61c5zi8pphjbcwxxjlgc34p61ic9-glibc-2.35-163-bin
2.5M    /nix/store/4pqv2mwdn88h7xvsm7a5zplrd8sxzvw0-glibc-2.35-163-dev
3.6M    /nix/store/4mxnw95jcm5a27qk60z7yc0gvxp42b9a-openssl-3.0.7
4.1M    /nix/store/mdck89nsfisflwjv6xv8ydj7dj0sj2pn-gcc-11.3.0-lib
5.9M    /nix/store/i38jcxrwa4fxk2b7acxircpi399kyixw-linux-headers-6.0
18M    /nix/store/4nlgxhb09sdr51nc9hdm8az5b08vzkgx-glibc-2.35-163
34M    /nix/store/46dhzhqqpwiinpzcwpyxj5nwljy5ff0h-chainweb
107M    /nix/store/1gf2flfqnpqbr1b4p4qz2f72y42bs56r-gcc-11.3.0
132M    /nix/store/480clkf74y2fqw8mjvp6vnz4sz6vzwx2-chainweb-exe-cwtool-2.18.1
132M    /nix/store/b82prm0kc08fs4izr2n9db4b41fsk1qg-chainweb-exe-chainweb-node-2.18.1
445M    total
```
If you look, the stripped copy is 34mb, yet it still depends on the unstripped version

Running `strings` and grepping for the cwtool sha:
```
❯ nix-shell -p --run "strings result/bin/cwtool | grep 480clkf74y2fqw8mjvp6vnz4sz6vzwx2
/nix/store/480clkf74y2fqw8mjvp6vnz4sz6vzwx2-chainweb-exe-cwtool-2.18.1/lib64:/nix/store/480clkf74y2fqw8mjvp6vnz4sz6vzwx2-chainweb-exe-cwtool-2.18.1/lib:/nix/store/026hln0aq1hyshaxsdvhg0kmcm6yf45r-zlib-1.2.13/lib:/nix/store/4mxnw95jcm5a27qk60z7yc0gvxp42b9a-openssl-3.0.7/lib:/nix/store/5056r5jmd1p22ji9wlh24gdj2zvhggfm-snappy-1.1.9/lib:/nix/store/70in66z4smj2kxs24jibnrj3r5q9gryp-gflags-2.2.2/lib:/nix/store/ysl6qj5r7nn63b16954dhk7x47r5yq7i-bzip2-1.0.8/lib:/nix/store/2rgwrp4zwia0ayhhxzw000cnjvr51pxa-lz4-1.9.4/lib:/nix/store/w10in9diaqrcqqxi5lg20n3q2jfpk6pq-zstd-1.5.2/lib:/nix/store/ijz81p08bp812q2bvv77lz9qpfzncibd-gmp-with-cxx-6.2.1/lib:/nix/store/664hlv5rfi4j1z1sa6jzj49sg7l2k03m-numactl-2.0.16/lib:/nix/store/xghl0l2f6byzqvy06ana2y4isqwa7blw-libffi-3.4.4/lib:/nix/store/4nlgxhb09sdr51nc9hdm8az5b08vzkgx-glibc-2.35-163/lib:/nix/store/mdck89nsfisflwjv6xv8ydj7dj0sj2pn-gcc-11.3.0-lib/lib
```
`ls`'ing the $out/lib path for cwtool:
```
❯ ls -lh /nix/store/480clkf74y2fqw8mjvp6vnz4sz6vzwx2-chainweb-exe-cwtool-2.18.1/lib
"/nix/store/480clkf74y2fqw8mjvp6vnz4sz6vzwx2-chainweb-exe-cwtool-2.18.1/lib": No such file or directory (os error 2)
```
So, the rpath includes a directory that never existed. Something is just blindly adding `$out/lib` to rpath, and nothing is running shrink.
This is easy enough to fix with `patchelf --shrink-rpath`:
```nix
    default = pkgs.runCommandCC "chainweb" {} ''
      mkdir -pv $out/bin
      cp ${flake.packages."chainweb:exe:chainweb-node"}/bin/chainweb-node $out/bin/chainweb-node
      cp ${flake.packages."chainweb:exe:cwtool"}/bin/cwtool $out/bin/cwtool
      chmod +w $out/bin/{cwtool,chainweb-node}
      $STRIP $out/bin/chainweb-node
      $STRIP $out/bin/cwtool
      patchelf --shrink-rpath $out/bin/{cwtool,chainweb-node}
    '';
```

Now you can see it shrinks a lot:
```
❯ du -hc --max=0 $(nix-store -qR ./result) | sort -h
35K    /nix/store/5056r5jmd1p22ji9wlh24gdj2zvhggfm-snappy-1.1.9
40K    /nix/store/xghl0l2f6byzqvy06ana2y4isqwa7blw-libffi-3.4.4
60K    /nix/store/ysl6qj5r7nn63b16954dhk7x47r5yq7i-bzip2-1.0.8
89K    /nix/store/026hln0aq1hyshaxsdvhg0kmcm6yf45r-zlib-1.2.13
155K    /nix/store/2rgwrp4zwia0ayhhxzw000cnjvr51pxa-lz4-1.9.4
232K    /nix/store/664hlv5rfi4j1z1sa6jzj49sg7l2k03m-numactl-2.0.16
271K    /nix/store/5mh5019jigj0k14rdnjam1xwk5avn1id-libidn2-2.3.2
493K    /nix/store/ijz81p08bp812q2bvv77lz9qpfzncibd-gmp-with-cxx-6.2.1
610K    /nix/store/70in66z4smj2kxs24jibnrj3r5q9gryp-gflags-2.2.2
634K    /nix/store/w10in9diaqrcqqxi5lg20n3q2jfpk6pq-zstd-1.5.2
920K    /nix/store/dsd5gz46hdbdk2rfdimqddhq6m8m8fqs-bash-5.1-p16
983K    /nix/store/34xlpp3j3vy7ksn09zh44f1c04w77khf-libunistring-1.0
3.6M    /nix/store/4mxnw95jcm5a27qk60z7yc0gvxp42b9a-openssl-3.0.7
4.1M    /nix/store/mdck89nsfisflwjv6xv8ydj7dj0sj2pn-gcc-11.3.0-lib
18M    /nix/store/4nlgxhb09sdr51nc9hdm8az5b08vzkgx-glibc-2.35-163
34M    /nix/store/5f9xlnbyrfmz7a506qjbph41ycd143km-chainweb
64M    total
```

**Comment by enobayram** [comment]
@chessai Nice work! Thanks for the detailed analysis and the default package optimizations.
